pkg/nats/topic.go

```diff
@@ -0,0 +1,39 @@
+package nats
+
+import (
+	"github.com/nats-io/nats.go"
+)
+
+type topicInterpreter struct {
+	js                nats.JetStreamManager
+	subjectCalculator SubjectCalculator
+}
+
+func newTopicInterpreter(js nats.JetStreamManager, formatter SubjectCalculator) *topicInterpreter {
+	if formatter == nil {
+		formatter = DefaultSubjectCalculator
+	}
+
+	return &topicInterpreter{
+		js:                js,
+		subjectCalculator: formatter,
+	}
+}
+
+func (b *topicInterpreter) ensureStream(topic string) error {
+	_, err := b.js.StreamInfo(topic)
+
+	if err != nil {
+		_, err = b.js.AddStream(&nats.StreamConfig{
+			Name:        topic,
+			Description: "",
+			Subjects:    b.subjectCalculator(topic).All(),
```

Comment:
could it make sense to make it configurable? for example by providing a closure that receive topic and returns  `nats.StreamConfig`?

Reply:
I like this idea, providing this level of control might fit nicely with a stripped down configuration in general (most of the config was just carried from v1 and some could work oddly - single durable name or queue group configured when a given subscriber creates multiple subscriptions to different topics is a good example).  The jetstream API has gotten so flexible that this is almost the only viable strategy to support provisioning like this I think.


To that end it might make sense to get rid of auto provisioning and lean into watermill's  SubscribeInitialize functionality too.  Provisioning on publish doesn't make a ton of sense to me but I mainly end up dealing with resources managed outside the app in production, haven't thought about it too deeply.